piet-wgsl/src/shaders/preprocess.rs

```diff
@@ -71,29 +87,53 @@ pub fn preprocess(
                             item.active = !item.active;
                         }
                     }
+                    let remainder = directive_start[directive_len..].trim();
+                    if !remainder.is_empty() {
+                        eprintln!("#else directives don't take an argument. `{remainder}` will not be in output (line {line_number})");
+                    }
+                    continue 'outer;
                 }
                 "endif" => {
                     if let None = stack.pop() {
                         eprintln!("Mismatched endif (line {line_number})");
                     }
+                    continue 'outer;
                 }
                 "import" => {
-                    let import = imports.get(arg);
+                    output.push_str(&line[..hash_index]);
+                    let directive_end = &directive_start[directive_len..];
+                    let import_name_start = if let Some(import_name_start) =
+                        directive_end.find(|c: char| !c.is_whitespace())
+                    {
+                        import_name_start
+                    } else {
+                        eprintln!("#import needs a non_whitespace argument (line {line_number})");
+                        continue 'outer;
+                    };
+                    let import_name_start = &directive_end[import_name_start..];
+                    let import_name_end_index = import_name_start
+                        .find(|c| !(matches!(c, '_') || c.is_alphanumeric()))
+                        .unwrap_or(import_name_start.len());
+                    let import_name = &import_name_start[..import_name_end_index];
+                    line = &import_name_start[import_name_end_index..];
+                    let import = imports.get(import_name);
                     if let Some(import) = import {
-                        output.push_str(&preprocess(import, defines, imports));
+                        if stack.last().map(|x| x.active).unwrap_or(true) {
+                            output.push_str(&preprocess(import, defines, imports));
+                        }
                     } else {
-                        eprintln!("Unkown import `{arg}` (line {line_number})");
+                        eprintln!("Unknown import `{import_name}` (line {line_number})");
                     }
+                    continue;
                 }
                 val => {
-                    eprintln!("Unknown preprocessor directive `{val}` (line {line_number})")
+                    eprintln!("Unknown preprocessor directive `{val}` (line {line_number})");
                 }
             }
-        } else {
-            if stack.last().map(|x| x.active).unwrap_or(true) {
-                output.push_str(line);
-                output.push('\n');
-            }
+        }
+        if stack.last().map(|x| x.active).unwrap_or(true) {
```

Comment:
Is this actually correct? It seems like a sequence of `#ifdef resolves_false`, `#ifdef resolves_true`, `inner`, `#endif`, `#endif` will actually cause inner to be included. My intuition (without having experimented) is that we want an `all` on the `active` field of all entries on the stack.

Reply:
Good point. Thanks